src/server/worldserver/IRCBridge/IRCBridge.cpp

```diff
@@ -0,0 +1,294 @@
+#include "Common.h"
+#include "Timer.h"
+#include "Config.h"
+#include "Log.h"
+#include "RelayHandler.h"
+#include "World.h"
+#include "IRCBridge.h"
+#include "IRCBridgeSocket.h"
+
+#include <boost/algorithm/string.hpp>
+#include <boost/asio/connect.hpp>
+#include <boost/asio/io_service.hpp>
+#include <boost/asio/ip/tcp.hpp>
+#include <functional>
+#include <thread>
+
+#define UPDATE_CONST 50
+
+IRCBridge::IRCBridge() : _ioService(nullptr), _strand(nullptr), _socket(nullptr), _status(IRCBRIDGESTATUS_IDLE), _relayType(RELAYTARGETTYPE_IRCBRIDGE), _reconnectTimer(0), _reconnectCounter(0), _active(false)
+{
+}
+
+IRCBridge::~IRCBridge()
+{
+    _status = IRCBRIDGESTATUS_SHUTDOWN;
+
+    if (_socket)
+        _socket->CloseSocket();
+
+    _active = false;
+
+    if (_thread.joinable())
+        _thread.join();
+
+    delete _strand;
+}
+
+IRCBridge* IRCBridge::instance()
+{
+    static IRCBridge instance;
+    return &instance;
+}
+
+void IRCBridge::Initialize(boost::asio::io_service * service)
+{
+    if (_active)
+        return;
+
+    _ioService = service;
+    _strand = new boost::asio::strand(*_ioService);
+    _thread = std::thread(std::bind(&IRCBridge::Run, this));
+
+    TC_LOG_DEBUG("IRCBridge", "IRCBridge::Initialize: thread initialized");
+}
+
+void IRCBridge::Run()
+{
+    if (!LoadConfigurations())
+        return;
+
+    _active = true;
+    while (_active)
+        ThreadLoop();
+}
+
+void IRCBridge::Stop()
+{
+    _status = IRCBRIDGESTATUS_SHUTDOWN;
+
+    sRelayHandler->Enable(_relayType, false);
+
+    if (_socket)
+        _socket->CloseSocket();
+
+    _active = false;
+
+    if (_thread.joinable())
```

Comment:
I'm pretty sure this is intentional, as this is in the destructor and is part of the shutdown code. It only joins on the thread if it is alive and not already joined.